### PR TITLE
SERVLET4_PUSH_TESTS-21780 #21780

### PR DESCRIPTION
**Pull request by edburns**
Tests for issue #21780 

	modified:   .gitignore
	modified:   main/appserver/tests/appserv-tests/devtests/web/build.xml
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/build.properties
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/build.xml
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/docroot/index.html
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/docroot/my.css
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/TestServlet.java
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/WebTest.java
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/build.properties
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/build.xml
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java
	new file:   main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/TestServlet.java

- Tests from before change to github.

### CODE REVIEW

**Comment by edburns** [comment]
@glassfishrobot Run CI tests please

**Comment by edburns** [comment]
http://gf-hudson.us.oracle.com/hudson/job/gf-remote-queue/344/

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) 2012-2017 Oracle and/or its affiliates. All rights reserved.
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java]: This should just be 2017 as it's a new file.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/build.xml [hunk 2]
```diff
@@ -0,0 +1,115 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!--
+
+    DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+    Copyright (c) 1997-2017 Oracle and/or its affiliates. All rights reserved.
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/build.xml]: The year should just be 2017.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+@WebServlet(urlPatterns="/foo/*")
+public class FooMethodServlet extends HttpServlet {
+        
+    @Override
+    protected void service(HttpServletRequest req, HttpServletResponse res)
+            throws IOException, ServletException {
+            String pathInfo = req.getPathInfo();
+            String method = req.getMethod();
+            if ("FOO".equals(method)) {
+                    System.out.println("debug: edburns: sending OK: pathInfo:" + pathInfo);
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java]: Probably want to remove this...

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/WebTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) 2012-2017 Oracle and/or its affiliates. All rights reserved.
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/WebTest.java]: Copyright year.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/build.xml [hunk 5]
```diff
@@ -0,0 +1,115 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!--
+
+    DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+    Copyright (c) 1997-2017 Oracle and/or its affiliates. All rights reserved.
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/build.xml]: Copyright year.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                    if ("/my.css".equals(pathInfo)) {
+                            res.getWriter().println("foo: bar");
+                            res.setStatus(HttpServletResponse.SC_OK);
+
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java]: Shouldn't this return after setting the status?

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                    }
+            }
+            if ("/my.css".equals(pathInfo)) {
+                    res.getWriter().println("foo: bar");
+                    res.setStatus(HttpServletResponse.SC_OK);
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java]: shouldn't this return after setting the status?

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                            res.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+                            return;
+                    }
+            }
+            if (pathInfo.contains("parameterMapFailure.js")) {
+                    Object result;
+                    if (null != (result = req.getParameter("ln"))) {
+                            if (!"resources".equals(result)) {
+                                    req.getServletContext().log("Parameter value incorrect.  Should be resources.  Is: " + result);
+                                    res.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java]: Shouldn't a return be here after setting the status?

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                            }
+                    } else {
+                            req.getServletContext().log("Should have ln parameter.  Does not.");
+                            res.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+
```
Comment (rlubke) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java]: Missing return?

**Review by ghost — COMMENTED** [review]
Hi Ed, can you comment on the returns?

**Review by ghost — DISMISSED** [review]

**Comment by edburns** [comment]
@glassfishrobot run CI tests please

**Comment by edburns** [comment]
http://gf-hudson.us.oracle.com/hudson/job/gf-remote-queue/356/

**Comment by shingwaichan** [comment]
I have the following comments on the change:
1. pushArbitraryMethod/servlet/test/TestServlet.java does not write the body to invoke FOO foo/my.css. 
2. pushArbitraryMethod/WebTest.java has comments and description for "push static file"
3. pushDynamic/WebTest.java has comments and description for "push static file"

It would be good to fix that.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java (hunk 3 above)
Comment (edburns) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java]: Thanks, removed it.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java (hunk 7 above)
Comment (edburns) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/servlet/test/FooMethodServlet.java]: Fixed.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java (hunk 9 above)
Comment (edburns) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushDynamic/servlet/test/NOTfacesServlet.java]: Fixed.

**Review by shingwaichan — CHANGES_REQUESTED** [review]
1. pushArbitraryMethod/WebTest.java, the following description is not correct : 
    stat.addDescription("Unit test for http2 push static file");
2. pushArbitraryMethod/servlet/test/TestServlet.java, we have the following: 
res.getWriter().println("<html><head><title>HTTP2 Test</title><link rel=\"stylesheet\" href=\"/foo/my.css\"></head><body>Hello</body></html>"); 
    The above html body will not access “foo/my.css” through the FOO method.  
    In other words, the pushed “foo/my.css” will not be accessed by a browser.

**Comment by edburns** [comment]
20170605-glassfish-pr-21786-pusharbitrarymethod

This image of chrome net-internals shows the test is effective in the browser as written.

**Comment by edburns** [comment]
I assert that no further change is needed to the pushArbitraryMethod test and request this review be approved.  

Thanks,

Ed

**Comment by shingwaichan** [comment]
The chrome net-internals shows the FOO request has arrived. My original change request is that the given http response body "need to" spawn a FOO request. (In other words, the client does not ask for the push promise here.)
Per discussion with Ed, he will update the WebTest.java to verify that the method of push promise is  "FOO" by using client API. And he will investigate the use of javascript to spawn FOO request issue later.

**Review by shingwaichan — DISMISSED** [review]

**Comment by edburns** [comment]
@glassfishrobot run ci tests please

**Comment by edburns** [comment]
@glassfishrobot run CI tests please

**Review by ghost — DISMISSED** [review]

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved.
+ *
+ * The contents of this file are subject to the terms of either the GNU
+ * General Public License Version 2 only ("GPL") or the Common Development
+ * and Distribution License("CDDL") (collectively, the "License").  You
+ * may not use this file except in compliance with the License.  You can
+ * obtain a copy of the License at
+ * https://glassfish.dev.java.net/public/CDDL+GPL_1_1.html
```
Comment (arindam-bandyopadhyay) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java]: This is old copyright header.Please update the copyright to the latest copyright header.

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/build.properties [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+    Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved.
+
+    The contents of this file are subject to the terms of either the GNU
+    General Public License Version 2 only ("GPL") or the Common Development
+    and Distribution License("CDDL") (collectively, the "License").  You
+    may not use this file except in compliance with the License.  You can
+    obtain a copy of the License at
+    https://glassfish.dev.java.net/public/CDDL+GPL_1_1.html
```
Comment (arindam-bandyopadhyay) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/build.properties]: This is old copyright header.Please update the copyright to the latest copyright header.

**Comment by shingwaichan** [comment]
To be consistence, we should update main/copyright/src/main/resources/copyright-exclude
 for the copyright rather than adding build.properties in copyright header.
Please see d8c544cfba9bd0dd5fb6b90e5e6943a9bbac3d05 as an example.

**Review by shingwaichan — APPROVED** [review]

File: main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java (hunk 10 above)
Comment (edburns) [inline main/appserver/tests/appserv-tests/devtests/web/servlet-4.0/pushArbitraryMethod/WebTest.java]: I removed the copyright statement and used the copyright-excludes.

**Comment by edburns** [comment]
@Glassfishrobot run CI tests please
